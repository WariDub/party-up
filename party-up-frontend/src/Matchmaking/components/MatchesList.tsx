import { Box, Button, Card, CardContent, GridList, GridListTile } from '@material-ui/core';
import React from 'react';
import { RouteComponentProps } from 'react-router-dom';
import Match from '../models/match.model';
import * as axios from 'axios';
import * as jwtDecode from 'jwt-decode';
import User from '../../Profile/models/user.model';
import { BACKEND_URL } from '../../globals';
import { JwtPayload } from '../../Auth/interfaces/jwt-payload.interface';
import { UsersService } from '../../Profile/users.service';

export interface MatchesListProps extends RouteComponentProps {
  matches: Match[];
}

export interface MatchesListState {
  user: User | null;
}

const MatchesList = class extends React.PureComponent<MatchesListProps, MatchesListState> {
  usersService = new UsersService();

  constructor(props: MatchesListProps) {
    super(props);

    this.state = {
      user: null,
    };
  }

  async componentDidMount() {
    await this.fetchAndSetUser();
  }

  render(): JSX.Element {
    const { matches } = this.props;
    const { user } = this.state;

    return (
      <Box p={2}>
        <GridList spacing={16} cellHeight="auto" cols={0}>
          {user
            ? matches.map((match) => (
                <GridListTile key={match.entry.user._id} onClick={this.handleMatchOnClick(match)}>
                  {this.renderMatch(match)}
                </GridListTile>
              ))
            : []}
        </GridList>
      </Box>
    );
  }

  renderMatch(match: Match): JSX.Element {
    const { user } = this.state;

    return (
      <Card>
        <CardContent>
          <h3>{match.entry.user.displayName}</h3>
          <h5>{`In-game: ${match.entry.identifier}`}</h5>
          <h5>{`${match.percentage}% match`}</h5>
          <Button
            variant="contained"
            color="primary"
            onClick={this.handleButtonOnClickToggleFriend(match)}
          >
            {user?.friends.includes(match.entry.user._id) ? 'Unfriend' : 'Add Friend'}
          </Button>
        </CardContent>
      </Card>
    );
  }

  handleMatchOnClick = (param: Match) => (event: any): void => {
    console.log('match on click');
  };

  handleButtonOnClickToggleFriend = (param: Match) => (event: any): void => {
    const { user } = this.state;
    if (!user) {
      return;
    }

    const addFriend = !user.friends.includes(param.entry.user._id);

    if (addFriend) {
      user.friends.push(param.entry.user._id);
      event.target.innerHTML = 'Unfriend';
      this.usersService.addFriend(user.username, param.entry.user._id);
    } else {
      user.friends = user.friends.filter((friend) => friend != param.entry.user._id);
      event.target.innerHTML = 'Add Friend';
      this.usersService.deleteFriend(user.username, param.entry.user._id);
    }
  };

  fetchAndSetUser = async (): Promise<void> => {
    const accessToken = localStorage.getItem('token');
    if (!accessToken) {
      return console.error('missing access token');
    }

    const payload: JwtPayload = jwtDecode.default(accessToken);

    try {
      const user = await this.usersService.getUser(payload.uid);
      if (user) {
        this.setState({ user });
      }
    } catch (e) {
      console.error(e);
    }
  };
};

export default MatchesList;
