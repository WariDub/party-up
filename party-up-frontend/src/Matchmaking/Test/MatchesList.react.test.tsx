import MatchesList from '../components/MatchesList';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { createBrowserHistory } from 'history';
import { matchPath } from 'react-router';
import Match from '../models/match.model';
import MatchmakingEntryModel from '../models/matchmaking-entry.model';
import User from '../../Profile/models/user.model';
import experience from '../../Matchmaking/enums/experience.enum';
import Gender from '../../Profile/enums/gender.enum';
import Genre from '../../Profile/enums/genre.enum';
import Role from '../../Profile/enums/role.enum';

describe('MatchestList', () => {
    test('It should mount', () => {
        let history = createBrowserHistory();
        history.push('/');

        let location = history.location;

        let match = matchPath('/',{
            path: '/',
            exact: true,
            strict: false
        });

        let user: User = {
            _id: '1',
            username: 'TestUsername',
            displayName: 'TestDisplayName',
            email: 'testEmail@testemail.com',
            avatarUrl: 'testURL',
            friends: ['TestFriend1', 'TestFriend2'],
            createdAt: 4/27/2021,
            age: 21,
            gender: Gender.MALE,
            favoriteGenre: Genre.HORROR,
            favoriteRole: Role.DPS
        }

        let entryMode: MatchmakingEntryModel = {
            user: user,
            experience: experience.BEGINNER,
            createdAt: 4/27/2021,
            identifier: 'TestIdentifier'
        }

        let matchymatch: Match = {
            entry: entryMode,
            percentage: 80
        }

        let component = shallow(<MatchesList history = { history } location = { location } matches = { [matchymatch] } match = { match }/>);

        expect(component.length).toBe(1);
    });
    
    test('It should mount', () => {
        let history = createBrowserHistory();
        history.push('/');

        let location = history.location;

        let match = matchPath('/MatchmakingPage',{
            path: '/MatchmakingPage',
            exact: true,
            strict: false
        });

        let user: User = {
            _id: '1',
            username: 'TestUsername',
            displayName: 'TestDisplayName',
            email: 'testEmail@testemail.com',
            avatarUrl: 'testURL',
            friends: ['TestFriend1', 'TestFriend2'],
            createdAt: 4/27/2021,
            age: 21,
            gender: Gender.MALE,
            favoriteGenre: Genre.HORROR,
            favoriteRole: Role.DPS
        }

        let entryMode: MatchmakingEntryModel = {
            user: user,
            experience: experience.BEGINNER,
            createdAt: 4/27/2021,
            identifier: 'TestIdentifier'
        }

        let matchymatch: Match = {
            entry: entryMode,
            percentage: 80
        }

        let matches: Match[] = [matchymatch];

        let component = renderer.create(<MatchesList history = { history } location = { location } matches = { matches } match = { match }/>).toJSON();

        expect(component).toMatchSnapshot();
    });
})