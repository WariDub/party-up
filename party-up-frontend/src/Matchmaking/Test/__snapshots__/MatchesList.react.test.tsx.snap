// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MatchestList It should mount 1`] = `
<div
  className="MuiBox-root MuiBox-root-1"
>
  <ul
    className="MuiGridList-root"
    style={
      Object {
        "margin": -4,
      }
    }
  >
    <li
      className="MuiGridListTile-root"
      onClick={[Function]}
      style={
        Object {
          "height": "auto",
          "padding": 4,
          "width": "Infinity%",
        }
      }
    >
      <div
        className="MuiGridListTile-tile"
      >
        <div
          className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            className="MuiCardContent-root"
          >
            <h3>
              TestDisplayName
            </h3>
            <h5>
              undefined
            </h5>
            <h5>
              80% match
            </h5>
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
